consensus/service/src/api/client_api_service.rs

```diff
@@ -308,8 +367,21 @@ impl ConsensusClientApi for ClientApiService {
                     ConsensusGrpcError::NotServing.into()
                 }
             } else {
-                self.handle_proposed_tx(msg)
-                    .or_else(ConsensusGrpcError::into)
+                let result = self.handle_proposed_tx(msg);
+                // The block present below rate-limits suspicious behavior.
+                if let Err(_err) = &result {
+                    let mut tracker = self.tracked_sessions.lock().expect("Mutex poisoned");
+                    let record = if let Some(record) = tracker.get_mut(&session_id) {
+                        record
+                    } else {
+                        tracker.put(session_id.clone(), ClientSessionTracking::new());
+                        tracker
+                            .get_mut(&session_id)
+                            .expect("Adding session-tracking record should be atomic.")
+                    };
+                    record.fail_tx_proposal(Instant::now(), self.config.tx_failure_window);
+                }
+                result.or_else(ConsensusGrpcError::into)
```

Comment:
:question: Thoughts on using a match statement here?
Something about seeing is_err logic and then in a separate block modifying that error.